aioesphomeapi/_frame_helper/base.py

```diff
@@ -64,13 +64,43 @@ def _set_ready_future_exception(self, exc: Exception) -> None:
         if not self._ready_future.done():
             self._ready_future.set_exception(exc)
 
-    def _read_exactly(self, length: int) -> bytearray | None:
+    def _append_or_replace_buffer(self, data: bytes) -> None:
+        """Append data to the buffer."""
+        if not self._buffer:
+            self._buffer = data
+            self._buffer_len = len(data)
+            return
+        # This is the expensive case since we have to create
+        # a new immutable bytes object each time, but since
+        # its so rare we do not want to optimize for it
+        self._buffer += data
+        self._buffer_len += len(data)
+
+    def _remove_packet_from_buffer(self) -> None:
+        """Remove the packet from the buffer."""
+        # Best case, the buffer is now emtpy
+        if self._buffer_len == self._pos:
+            self._buffer = None
+            self._buffer_len = 0
+            return
+        if TYPE_CHECKING:
+            assert self._buffer is not None
+        # This is the expensive case since we have to create
+        # a new immutable bytes object each time, but since
+        # its so rare we do not want to optimize for it
+        end_of_frame_pos = self._pos
+        self._buffer_len -= end_of_frame_pos
+        self._buffer = self._buffer[end_of_frame_pos:]
```

Comment:
This happens when the event loop is a bit behind but its not that bad since it means we only slice once per packet